Documentation/reference-guides/xfrm/index.rst

```diff
@@ -0,0 +1,301 @@
+
+.. code-block:: text
+
+    src 0.0.0.0/0 dst 10.56.0.0/24 
+        dir in priority 0 
+        mark 0xd00/0xf00 
+        tmpl src 10.56.1.238 dst 10.56.0.17
+            proto esp reqid 1 mode tunnel
+
+Note that when using tunnel mode as is the case here, we should always see XFRM states matching the template of XFRM OUT policies. That is because, on egress, the states are matched after the template is applied. The IP addresses, the SPI, the protocol, the mode, and the reqid should all match between the XFRM state and the template in that case.
+
+
+XFRM Packet Flows
+=================
+
+IPsec and XFRM are represented in the usual Linux networking diagram:
+
+.. image:: /images/netfilter-with-xfrm-fix.png
+    :align: center
+
+Egress Packet Flow
+------------------
+
+On egress, packets will first hit the "XFRM lookup" block. At this point, a lookup is performed against the XFRM OUT policies. If a match is found, the packet goes to the "XFRM encode" block, any template is applied (ex. encapsulation), and the packet is then matched against XFRM states. If a state is found, its information is used to encrypt the packet.
+
+The encrypted packet will then navigate again through the OUTPUT and POSTROUTING chains.
+
+Ingress Packet Flow
+-------------------
+
+On ingress, encrypted packets will hit the "XFRM lookup" after they navigate through the INPUT chain.
+
+In tunnel mode, encrypted packets will typically have one of the server's IP addresses as the outer destination address, so they should automatically be routed through the INPUT chain. If not, it may be necessary to add IP routes to redirect packets to the INPUT chain. As an example, Cilium identifies IPsec traffic on tc-bpf ingress and marks them with a special value which is then used to reroute those packets to the INPUT chain.
+
+At the "XFRM lookup", if packets match an XFRM state, they will be decoded (i.e., decapsulated and decrypted) using the state's information. The match is based on the source & destination addresses, the mark, the SPI, and the protocol. In case of any decode error (ex. wrong key), the packet is dropped and an error counter is increased.
+
+Note that an XFRM policy matching the packet isn't required for the decoding to happen, but it may be required for the packet to proceed through the FORWARD chain. An XFRM policy with an optional template (i.e., ``level use``) is also required if no state nor socket matches it (ex. userspace proxy redirects in Cilium).
+
+After a packet is decoded, it is recirculated in the stack. More specifically packets are recirculated before the tc layer, such that they are visible on the tc-bpf hook a second time (once before decryption, once after). Note that the famous Netfilter packet flow diagram is incorrect on that point (the image above has been fixed). The packet mark is preserved when recirculated, so it's possible to identify and trace packets that have been decrypted and recirculated.
```

Comment:
Maybe worth saying explicitly "the decrypted packet is recirculated as if coming from the interface it was received on".